File: app/src/main/java/com/aprivate/sean/ohcg/ScoreBoardItemAdapter.java
package com.aprivate.sean.ohcg;

import android.app.Activity;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import java.util.Collections;
import java.util.List;

public class ScoreBoardItemAdapter extends BaseAdapter {

    private Context scoreBoardContext;
    private List<ScoreBoardItem> scoreBoardItems;

    public ScoreBoardItemAdapter(Context context, List<ScoreBoardItem> items) {
        scoreBoardItems = items;
        scoreBoardContext = context;
    }

    public void add(ScoreBoardItem record) {
        scoreBoardItems.add(record);
        notifyDataSetChanged();
    }

    public void establishRankings(List<ScoreBoardItem> players) {
        Collections.sort(players, (player1, player2) -> player1.getCurrentPoints() < player2.getCurrentPoints() ? 1 : player1.getCurrentPoints() == player2.getCurrentPoints() ? 0 : -1);
        int rank = 1;
        for (ScoreBoardItem player : players){
            player.setBoardRank(rank++);
        }
    }

    public void update(ScoreBoardItem updateRecord){
        for (int i = 0; i < scoreBoardItems.size(); i++) {
            if(scoreBoardItems.get(i).getIdNumber() == updateRecord.getIdNumber()){
                scoreBoardItems.set(i, updateRecord);
                break;
            }
        }
        notifyDataSetChanged();
    }

    @Override
    public int getCount() {
        return scoreBoardItems.size();
    }
    @Override
    public Object getItem(int i) {
        return scoreBoardItems.get(i);
    }

    public List<ScoreBoardItem> getScoreBoardItems() {
        return scoreBoardItems;
    }


    @Override
    public long getItemId(int i) {
        return i;
    }

    @Override
    public View getView(int i, View view, ViewGroup viewGroup) {
        RecordViewHolder holder;

        if(i == 0){
            establishRankings(scoreBoardItems);
        }

        if (view ==null){
            LayoutInflater recordInflater = (LayoutInflater)
                    scoreBoardContext.getSystemService(Activity.LAYOUT_INFLATER_SERVICE);
            view = recordInflater.inflate(R.layout.score_board, null);

            holder = new RecordViewHolder();
            holder.Name = (TextView) view.findViewById(R.id.name);
            holder.CurrentBid = (TextView) view.findViewById(R.id.currentBid);
            holder.TotalPts = (TextView) view.findViewById(R.id.totalPoints);
            holder.Rank = (TextView) view.findViewById(R.id.rank);
            holder.Order = view.findViewById(R.id.order);
            view.setTag(holder);
        }else {
            holder = (RecordViewHolder) view.getTag();
        }

        ScoreBoardItem scoreBoardItem = (ScoreBoardItem) getItem(i);
        holder.Name.setText(scoreBoardItem.getPlayerName());
        holder.CurrentBid.setText(String.valueOf(scoreBoardItem.getCurrentBid()));
        holder.TotalPts.setText(String.valueOf(scoreBoardItem.getCurrentPoints()));
        holder.Rank.setText(String.valueOf(scoreBoardItem.getBoardRank()));
        holder.Order.setText(String.valueOf(scoreBoardItem.getOrder()));
        return view;
    }

    public void updateOrderAndPoints(ScoreBoardItem player, int newOrderNumber, int newPoints) {
        int previousOrder = player.getOrder();
        for (ScoreBoardItem item : scoreBoardItems) {
            if(item.getIdNumber() == player.getIdNumber()){
                item.setOrder(newOrderNumber);
                item.setCurrentPoints(newPoints);
            }else if(item.getOrder() < previousOrder && item.getOrder() >= newOrderNumber){
                item.setOrder(item.getOrder() + 1);
                update(item);
            } else if(item.getOrder() > previousOrder && item.getOrder() <= newOrderNumber){
                item.setOrder(item.getOrder() - 1);
            }else{
                continue;
            }
            update(item);
        }
    }

    private static class RecordViewHolder {

        public TextView Name;
        public TextView CurrentBid;
        public TextView TotalPts;
        public TextView Rank;
        public TextView Order;
    }
}

File: app/src/main/java/com/aprivate/sean/ohcg/ScoreBoardItemAdapter.kt
package com.aprivate.sean.ohcg

import android.app.Activity
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.TextView

import java.util.Collections

class ScoreBoardItemAdapter(private val scoreBoardContext: Context, val scoreBoardItems: MutableList<ScoreBoardItem>) : BaseAdapter() {

    fun add(record: ScoreBoardItem) {
        scoreBoardItems.add(record)
        notifyDataSetChanged()
    }

    fun establishRankings(players: List<ScoreBoardItem>) {
        Collections.sort(players) { player1, player2 -> if (player1.currentPoints < player2.currentPoints) 1 else if (player1.currentPoints == player2.currentPoints) 0 else -1 }
        var rank = 1
        for (player in players) {
            player.boardRank = rank++
        }
    }

    fun update(updateRecord: ScoreBoardItem) {
        for (i in scoreBoardItems.indices) {
            if (scoreBoardItems[i].idNumber == updateRecord.idNumber) {
                scoreBoardItems[i] = updateRecord
                break
            }
        }
        notifyDataSetChanged()
    }

    override fun getCount(): Int {
        return scoreBoardItems.size
    }

    override fun getItem(i: Int): Any {
        return scoreBoardItems[i]
    }

//    fun getScoreBoardItems(): List<ScoreBoardItem> {
//        return scoreBoardItems
//    }


    override fun getItemId(i: Int): Long {
        return i.toLong()
    }

    override fun getView(i: Int, view: View?, viewGroup: ViewGroup): View {
        var view = view
        val holder: RecordViewHolder

        if (i == 0) {
            establishRankings(scoreBoardItems)
        }

        if (view == null) {
            val recordInflater = scoreBoardContext.getSystemService(Activity.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            view = recordInflater.inflate(R.layout.score_board, null)

            holder = RecordViewHolder()
            holder.Name = view!!.findViewById<View>(R.id.name) as TextView
            holder.CurrentBid = view.findViewById<View>(R.id.currentBid) as TextView
            holder.TotalPts = view.findViewById<View>(R.id.totalPoints) as TextView
            holder.Rank = view.findViewById<View>(R.id.rank) as TextView
            holder.Order = view.findViewById(R.id.order)
            view.tag = holder
        } else {
            holder = view.tag as RecordViewHolder
        }

        val scoreBoardItem = getItem(i) as ScoreBoardItem
        holder.Name!!.text = scoreBoardItem.playerName
        holder.CurrentBid!!.text = scoreBoardItem.currentBid.toString()
        holder.TotalPts!!.text = scoreBoardItem.currentPoints.toString()
        holder.Rank!!.text = scoreBoardItem.boardRank.toString()
        holder.Order!!.text = scoreBoardItem.order.toString()
        return view
    }

    fun updateOrderAndPoints(player: ScoreBoardItem, newOrderNumber: Int, newPoints: Int, activeStatus: Boolean) {
        val previousOrder = player.order
        for (item in scoreBoardItems) {
            if (item.idNumber == player.idNumber) {
                item.order = newOrderNumber
                item.currentPoints = newPoints
                item.isActiveInGame = activeStatus
            } else if (item.order < previousOrder && item.order >= newOrderNumber) {
                item.order = item.order + 1
                update(item)
            } else if (item.order > previousOrder && item.order <= newOrderNumber) {
                item.order = item.order - 1
            } else {
                continue
            }
            update(item)
        }
    }

    private class RecordViewHolder {

        var Name: TextView? = null
        var CurrentBid: TextView? = null
        var TotalPts: TextView? = null
        var Rank: TextView? = null
        var Order: TextView? = null
    }
}
